src/jupyterlabcontroller/services/lab.py

```diff
@@ -0,0 +1,659 @@
+                mount_path="/opt/lsst/software/jupyterlab/runtime",
+                name=f"nb-{self.username}-runtime",
+                read_only=True,
+            ),
+        )
+        return runtime_vol
+
+    async def build_volumes(self) -> List[LabVolumeContainer]:
+        """This stitches together the Volume and VolumeMount definitions
+        from each of our sources.
+        """
+        # Begin with the /tmp empty_dir
+        vols: List[LabVolumeContainer] = []
+        lab_config_vols = await self.build_lab_config_volumes(
+            self.lab_config.volumes
+        )
+        vols.extend(lab_config_vols)
+        nss_vols = await self.build_nss_volumes()
+        vols.extend(nss_vols)
+        cm_vols = await self.build_cm_volumes()
+        vols.extend(cm_vols)
+        secret_vol = await self.build_secret_volume()
+        vols.append(secret_vol)
+        env_vol = await self.build_env_volume()
+        vols.append(env_vol)
+        tmp_vol = await self.build_tmp_volume()
+        vols.append(tmp_vol)
+        runtime_vol = await self.build_runtime_volume()
+        vols.append(runtime_vol)
+        return vols
+
+    async def build_init_ctrs(self) -> List[Container]:
+        init_ctrs: List[Container] = []
+        ic_volumes: List[LabVolumeContainer] = []
+        for ic in self.lab_config.initcontainers:
+            if ic.volumes is not None:
+                ic_volumes = await self.build_lab_config_volumes(ic.volumes)
+            ic_vol_mounts = [x.volume_mount for x in ic_volumes]
+            ic_sec_ctx = (
+                SecurityContext(
+                    run_as_non_root=True,
+                    run_as_user=1000,
+                    allow_privilege_escalation=False,
+                ),
+            )
+            if ic.privileged:
+                ic_sec_ctx = SecurityContext(
+                    run_as_non_root=False,
+                    run_as_user=0,
+                    allow_privilege_escalation=True,
+                )
+            ctr = Container(
+                name=ic.name,
+                image=ic.image,
+                security_context=ic_sec_ctx,
+                volume_mounts=ic_vol_mounts,
+            )
+            self.logger.debug(f"Added init container {ctr}")
+            init_ctrs.append(ctr)
+        return init_ctrs
+
+    async def build_pod_spec(self, user: UserInfo) -> PodSpec:
+        vol_recs = await self.build_volumes()
+        volumes = [x.volume for x in vol_recs]
+        vol_mounts = [x.volume_mount for x in vol_recs]
+        init_ctrs = await self.build_init_ctrs()
+        nb_ctr = Container(
+            name="notebook",
+            args=["/opt/lsst/software/jupyterlab/runlab.sh"],
+            env_from=EnvFromSource(
+                config_map_ref=ConfigMapEnvSource(
+                    name=f"nb-{self.username}-env"
+                )
+            ),
+            image=self.lab.options.image,
+            image_pull_policy="Always",
```

Comment:
It's historically been Always because our image tags are not immutable: we can build a different container and slap an existing tag on it, and have done that in a few cases when weeklies have gone out bad and later been fixed.